wsnet/cache.go

```diff
@@ -0,0 +1,154 @@
+package wsnet
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+)
+
+// dialerFunc is used to reference a dialer returned for caching.
+type dialerFunc func(ctx context.Context, key string) (*Dialer, error)
+
+// DialCache constructs a new DialerCache.
+// The cache clears connections that:
+// 1. Are older than the TTL and have no active user-created connections.
+// 2. Have been closed.
+func DialCache(ttl time.Duration, dialer dialerFunc) *DialerCache {
+	dc := &DialerCache{
+		ttl:         ttl,
+		dialerFunc:  dialer,
+		closed:      make(chan struct{}),
+		flightGroup: &singleflight.Group{},
+		mut:         sync.RWMutex{},
+		dialers:     make(map[string]*Dialer),
+		atime:       make(map[string]time.Time),
+	}
+	go dc.init()
+	return dc
+}
+
+type DialerCache struct {
+	dialerFunc  dialerFunc
+	ttl         time.Duration
+	flightGroup *singleflight.Group
+
+	closed  chan struct{}
+	mut     sync.RWMutex
+	dialers map[string]*Dialer
+	atime   map[string]time.Time
+}
+
+// init starts the ticker for evicting connections.
+func (d *DialerCache) init() {
+	ticker := time.NewTicker(time.Second * 30)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.closed:
+			return
+		case <-ticker.C:
+			d.evict()
+		}
+	}
+}
+
+// evict removes lost/broken/expired connections from the cache.
+func (d *DialerCache) evict() {
+	var wg sync.WaitGroup
+	d.mut.RLock()
+	for key, dialer := range d.dialers {
+		wg.Add(1)
+		key := key
+		dialer := dialer
+		go func() {
+			defer wg.Done()
+
+			evict := false
+			select {
+			case <-dialer.Closed():
+				evict = true
+			default:
+			}
```

Comment:
```suggestion
```

Reply:
Why remove this? We want to evict closed connections.

Reply:
oops I thought this was d.Closed my bad